cmake/init-submodule.cmake

```diff
@@ -0,0 +1,40 @@
+# credit: https://cliutils.gitlab.io/modern-cmake/chapters/projects/submodule.html
+
+find_package(Git QUIET)
+
+# NOTE: cannot use PROJECT_SOURCE_DIR here because we need to run this *before*
+#       `project()` is called in order to ensure vcpkg is available.
+
+if(GIT_FOUND AND EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/.git")
+    # Update submodules as needed
+    if(_TILEDB_CMAKE_INIT_GIT_SUBMODULES)
+        message(STATUS "Update submodules")
+        # TODO check pinned version here?
+        execute_process(COMMAND ${GIT_EXECUTABLE} submodule update --init --recursive
+                        WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
+                        RESULT_VARIABLE GIT_SUBMOD_RESULT
+                        ECHO_OUTPUT_VARIABLE
+                        COMMAND_ECHO STDOUT)
+        if(NOT GIT_SUBMOD_RESULT EQUAL "0")
+            message(FATAL_ERROR "git submodule update --init --recursive failed with ${GIT_SUBMOD_RESULT}, please checkout submodules")
+        endif()
+    endif()
+endif()
+
+if(NOT EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/external/vcpkg/LICENSE.txt")
+    message(FATAL_ERROR "The submodules were not downloaded! _TILEDB_CMAKE_INIT_GIT_SUBMODULES"
+                        " was turned off or failed. Please update submodules and try again.")
+endif()
+
+#############
+# vcpkg setup
+
+if (TILEDB_GCS)
+    # TODO our_vcpkg_enable
+    list(APPEND VCPKG_MANIFEST_FEATURES "gcs")
+
+    # This must be set *before* `project()`
+    # TODO make sure this works with user-specified toolchain
+    set(CMAKE_TOOLCHAIN_FILE "${CMAKE_CURRENT_SOURCE_DIR}/external/vcpkg/scripts/buildsystems/vcpkg.cmake"
```

Comment:
CMake's documentation says that `CMAKE_TOOLCHAIN_PATH` will be read early in the CMake run. I wonder if this place is early enough.

How about we create a `CMakePresets.json` file and set the toolchain file there? We would need it either way sooner or later. It will assume though that the vcpkg submodule is cloned.

Reply:
If you put a `message(FATAL_ERROR ...)` in the toolchain file, you can see where it is evaluated. Also:

> All vcpkg-affecting variables must be defined before the first project() directive

https://learn.microsoft.com/en-us/vcpkg/users/buildsystems/cmake-integration#settings-reference

I'm fine with a presets file in the future.